src/blocks/libvirt.rs

```
@@ -0,0 +1,167 @@
+#[serde(deny_unknown_fields, default)]
+pub struct LibvirtConfig {
+    /// Update interval in seconds
+    #[serde(deserialize_with = "deserialize_duration")]
+    pub interval: Duration,
+
+    /// Format override
+    pub format: FormatTemplate,
+
+    /// URL to QEMU
+    pub qemu_url: String,
+}
+
+impl Default for LibvirtConfig {
+    fn default() -> Self {
+        Self {
+            interval: Duration::from_secs(5),
+            format: FormatTemplate::default(),
+            qemu_url: "qemu:///system".to_string(),
+        }
+    }
+}
+
+impl ConfigBlock for Libvirt {
+    type Config = LibvirtConfig;
+
+    fn new(
+        id: usize,
+        block_config: Self::Config,
+        shared_config: SharedConfig,
+        _: Sender<Task>,
+    ) -> Result<Self> {
+        let text = TextWidget::new(id, 0, shared_config)
+            .with_text("vms")
+            .with_icon("virtual-machine")?;
+        Ok(Libvirt {
+            id,
+            text,
+            format: block_config.format.with_default("{running}")?,
+            update_interval: block_config.interval,
+            qemu_conn: Connect::open_read_only(&block_config.qemu_url)
+                .expect("could not connect to libvirtd"),
+        })
+    }
+}
+
+impl Block for Libvirt {
+    fn update(&mut self) -> Result<Option<Update>> {
+        if !self
+            .qemu_conn
+            .is_alive()
+            .expect("unrecoverable error with qemu object")
+        {
+            self.qemu_conn = Connect::open_read_only(
+                &self
+                    .qemu_conn
+                    .get_uri()
+                    .expect("could not get URI from the currently connected QEMU object"),
+            )
+            .expect("could not re-connect to libvirtd")
+        };
+
+        let mut paused: i64 = 0;
+        match self.qemu_conn.list_all_domains(1 << 5) {
+            Ok(d) => paused = d.len() as i64,
+            Err(e) => eprintln!("{}", e),
```

Perhaps it would be better to propagate the error instead of logging it?